### ISSUE TITLE
Bounds on associated types still ignored

### ISSUE THREAD

**Issue by bgamari** [rust-lang/rust#21636]
It seems #20890 may not be entirely resolved. Take this example,
```rust
pub trait BitIter {
    type Iter: Iterator<Item=bool>;
    fn bit_iter(self) -> <Self as BitIter>::Iter;
}

fn hi<T>(arg: T) where 
    T: BitIter,
    //<T as BitIter>::Iter: Iterator<Item=bool>
{
    for i in arg.bit_iter() {}
}
```

`rustc 1.0.0-dev (2a51ca0a4 2015-01-25 10:42:41 -0500)` fails to compile this with,
```
hi.rs:9:18: 9:28 error: type mismatch resolving `<<T as BitIter>::Iter as core::iter::Iterator>::Item == bool`:
 expected associated type,
    found bool [E0271]
hi.rs:9     for i in arg.bit_iter() {}
                         ^~~~~~~~~~
```

Unless, of course, I add the bound.

**Comment by huonw** [comment]
cc @nikomatsakis

**Comment by nikomatsakis** [comment]
I think the problem is because we are required to resolve the projection (that is, the result of `T::Iter::Item`) based on what is declared in the trait, and we don't currently have any logic for doing that.

**Comment by carllerche** [comment]
cc me

**Comment by nikomatsakis** [comment]
Fix in #21702

**Comment by bgamari** [comment]
Confirmed to be fixed.
